pkg/archive/archive.go

```diff
@@ -530,7 +550,8 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 		}
 
 	case tar.TypeLink:
-		targetPath := filepath.Join(extractDir, hdr.Linkname)
+		lpath := adjustHardLink(hdr.Linkname, hdr.Name)
+		targetPath := filepath.Join(extractDir, lpath)
```

Comment:
@keloyang I'm not sure I understand what you're saying. The problem is that when you generate the archive the `hdr.Linkname`s are not properly translated to the target path (they are correctly computed in the source but not in the destination). If we fix the archive generation there's no need to change clients, and you avoid having this on-the-fly rewriting (that makes me quite nervous).